File: 14.Dynamic Programming/Minimum Path Sum.java
package Week14_Dynamic_Programming;

public class Day1_Minimum_Path_Sum {
    public static int minPathSum(int[][] grid) {
    	int answer = 0;

    	int[][] dp = new int[grid.length][grid[0].length];
    	
    	dp = grid;
//    	for(int i = 0 ; i < grid.length ; i++) {
//    		for(int j = 0 ; j < grid[i].length ; j++) {
//    			System.out.print(dp[i][j] + " ");
//    		}
//    		System.out.println();
//    	}
    	
    	for(int i = 0 ; i < grid.length ; i++) {
    		for(int j = 0 ; j < grid[i].length ; j++) {
    			
    			if(i == 0 && j != 0) {
    				dp[i][j] +=  dp[i][j-1];
    			}
    			else if(j == 0 && i != 0) {
    				dp[i][j] += dp[i-1][j];
    			}
    			
    			if(i != 0  && j != 0) {
    				dp[i][j] = Math.min(dp[i-1][j] + dp[i][j] , dp[i][j-1] + dp[i][j]);
    				
    			}
    		}
    	}
//    	System.out.println();
//    	for(int i = 0 ; i < grid.length ; i++) {
//    		for(int j = 0 ; j < grid[i].length ; j++) {
//    			System.out.print(dp[i][j] + " ");
//    		}
//    		System.out.println();
//    	}
    	return dp[grid.length-1][grid[0].length-1];
    }
 
	public static void main(String[] args) {
//		int[][] grid = { {1,3,1}, {1,5,1}, {4,2,1} };
		int[][] grid = { {1,2,3}, {4,5,6}};
		System.out.println(minPathSum(grid));

	}

}

File: 14.Dynamic Programming/Minimum Path Sum.py
from typing import *

class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        ans = []
        m = len(grid)
        n = len(grid[0])

        for i in range(m):
            ans.append([])
            for j in range(n):
                ans[i].append(0)

        ans[0][0] = grid[0][0]

        for i in range(1, m):
            ans[i][0] = ans[i - 1][0] + grid[i][0]

        for j in range(1, n):
            ans[0][j] = ans[0][j - 1] + grid[0][j]

        for i in range(1, m):
            for j in range(1, n):
                ans[i][j] = min(ans[i - 1][j], ans[i][j - 1]) + grid[i][j]

        return ans[m - 1][n - 1]

sol = Solution()
grid = [[1,3,1],[1,5,1],[4,2,1]]
print(sol.minPathSum(grid))